### PR TITLE
Fix ability to activate next/previous pane in a dock

### LINKED ISSUE #14442: Uncaught TypeError: this.getActivePaneContainer(...).activateNextPane is not a function
[Enter steps to reproduce:]

1. Open a dock (I used the sidebar, but this works in any dock)
2. Trigger `window:focus-next-pane`

expected:
EITHER
1) Next pane/dock is selected
2) Nothing happens, as dock only has one pane

actual:
The below error happens.

**Atom**: 1.17.0 x64
**Electron**: 1.3.15
**OS**: Mac OS X 10.12.4
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: this.getActivePaneContainer(...).activateNextPane is not a function

```
At /Applications/Atom.app/Contents/Resources/app/src/workspace.js:1559

TypeError: this.getActivePaneContainer(...).activateNextPane is not a function
    at Workspace.activateNextPane (/Applications/Atom.app/Contents/Resources/app/src/workspace.js:1559:1)
    at atom-workspace.window:focus-next-pane (/Applications/Atom.app/Contents/Resources/app/src/register-default-commands.js:168:1)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:265:1)
    at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:1
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:610:1)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:401:1)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:100:1)
    at HTMLDocument.<anonymous> (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:3:1)
```

### Commands

```
  2x -0:51.1.0 window:toggle-left-dock (input.hidden-input)
  3x -0:45.6.0 pane:show-previous-item (input.hidden-input)
  6x -0:40.7.0 pane:show-next-item (input.hidden-input)
  5x -0:39.5.0 pane:show-previous-item (input.hidden-input)
     -0:38.4.0 pane:show-next-item (input.hidden-input)
     -0:36.3.0 window:toggle-left-dock (input.hidden-input)
     -0:33.7.0 core:move-right (input.hidden-input)
     -0:33.7.0 core:move-down (input.hidden-input)
     -0:33 core:move-left (input.hidden-input)
     -0:30.6.0 command-palette:toggle (input.hidden-input)
     -0:28 core:backspace (input.hidden-input)
     -0:26.6.0 editor:consolidate-selections (input.hidden-input)
     -0:26.6.0 core:cancel (input.hidden-input)
     -0:25.5.0 window:focus-next-pane (input.hidden-input)
  4x -0:24.4.0 core:move-down (input.hidden-input)
  3x -0:18.8.0 window:focus-next-pane (div.tool-panel.tree-view)
```

### Non-Core Packages

```
auto-update-packages 1.0.1 
autocomplete-haskell 0.7.2 
busy-signal 1.4.1 
file-icons 2.1.5 
fonts 2.1.0 
git-blame 1.0.2 
intentions 1.1.2 
jslint 1.5.1 
language-babel 2.61.0 
language-gherkin 1.0.4 
language-haskell 1.12.1 
linter 2.1.4 
linter-eslint 8.1.7 
linter-ui-default 1.5.0 
lisp-paredit 0.5.5 
nyancat 0.14.0 
sort-lines 0.14.0 
swackets 0.26.0 
```

### PR DESCRIPTION
**Pull request by jasonrudolph**
Fixes #14442.

### Description of the Change

Prior to this change, when a dock had focus, triggering `window:focus-next-pane` or `window:focus-previous-pane` would throw an error:

```
At /Applications/Atom.app/Contents/Resources/app/src/workspace.js:1559

TypeError: this.getActivePaneContainer(...).activateNextPane is not a function
    at Workspace.activateNextPane (/Applications/Atom.app/Contents/Resources/app/src/workspace.js:1559:1)
    at atom-workspace.window:focus-next-pane (/Applications/Atom.app/Contents/Resources/app/src/register-default-commands.js:168:1)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app/src/command-registry.js:265:1)
    at /Applications/Atom.app/Contents/Resources/app/src/command-registry.js:3:1
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:610:1)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app/node_modules/atom-keymap/lib/keymap-manager.js:401:1)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:100:1)
    at HTMLDocument.<anonymous> (/Applications/Atom.app/Contents/Resources/app/src/window-event-handler.js:3:1)
```

[`Workspace.activateNextPane()` calls `this.getActivePaneContainer().activateNextPane()`](https://github.com/atom/atom/blob/ae9f24602d80cb15fe0edc15c1aae37d44ee786c/src/workspace.js#L1361). `this.getActivePaneContainer()` returns either a `WorkspaceCenter` (if the workspace center is the active pane container) or a `Dock` (if a dock is the active pane container). `WorkspaceCenter` has an [`activateNextPane`](https://github.com/atom/atom/blob/8ed16cea43cfd3f3b04b51ed5851892944ac7f19/src/workspace-center.js#L290-L298) function, but `Dock` does not. Therefore, calling `Workspace.activateNextPane()` results in an error when a `Dock` is the active pane container. (The same goes for `activatePreviousPane`.)

This change resolves this issue by [adding `activateNextPane` and `activatePreviousPane` functions to `Dock`](https://github.com/atom/atom/blob/ae9f24602d80cb15fe0edc15c1aae37d44ee786c/src/dock.js#L629-L637).

With this change in place, triggering `window:focus-next-pane` or `window:focus-previous-pane` in a dock will focus the dock's next or previous pane respectively. If the dock has only one pane, then that pane will remain focused. In other words, these two commands should provide the same behavior regardless of whether they're executed in a dock or in the workspace center.

### Alternate Designs

None.

### Why Should This Be In Core?

It's a fix for an existing bug in atom/atom: #14442

### Benefits

Fixes #14442.

### Possible Drawbacks

From a functionality perspective, I'm not aware of any drawbacks.

From a code perspective:
- It seems like there's an interface that we expect both `Dock` and `WorkspaceCenter` to implement, but I don't see that interface explicitly described or enforced anywhere. I _think_ that lack of explicit interface is a cause of the bug seen in #14442. There may be other bugs that are lurking due to code that implicitly expects a `WorkspaceCenter` and a `Dock` to provide a common interface.
- [`Dock`](https://github.com/atom/atom/blob/8ed16cea43cfd3f3b04b51ed5851892944ac7f19/src/dock.js) and [`WorkspaceCenter`](https://github.com/atom/atom/blob/8ed16cea43cfd3f3b04b51ed5851892944ac7f19/src/workspace-center.js) have quite a bit of duplicate code ([side-by-side screenshot](https://cloud.githubusercontent.com/assets/2988/26175132/a7d9c034-3b1f-11e7-9586-f738646b7440.png)), and this PR adds even more. Perhaps a future refactoring could reduce this duplication.

### Applicable Issues

#14442

### Demo

https://github.com/atom/command-palette/issues/79#issue-227125354 shows the behavior prior to this change. The gif below shows the behavior after this change:

demo

### CODE REVIEW

File: spec/dock-spec.js [hunk 1]
```diff
@@ -62,6 +62,74 @@ describe('Dock', () => {
...
+        jasmine.attachToDOM(atom.workspace.getElement())
+        const dock = atom.workspace.getLeftDock()
+
+        expect(dock.getPanes().length).toBe(1)
+        const pane = dock.getPanes()[0]
+        expect(pane.isActive()).toBe(true)
+        dock.activateNextPane()
+        expect(pane.isActive()).toBe(true)
+      })
+    })
```
Comment (jasonrudolph) [inline spec/dock-spec.js]: @lee-dohm: In addition to the scenarios above where a dock has one pane or multiple panes, you and I discussed adding a test for the scenario where the dock has zero panes. I haven't been able to set up that scenario in a test. I thought something like this would work:

```js
describe('when the dock has no panes', () => {
  it('has no effect', async () => {
    jasmine.attachToDOM(atom.workspace.getElement())
    const dock = atom.workspace.getLeftDock()
    await dock.destroyActivePane()

    expect(dock.getPanes().length).toBe(0)
    dock.activateNextPane()
    expect(dock.getPanes().length).toBe(0)
  })
})
```

But upon closer inspect, it looks like Atom [prevents destruction of the last pane in a container](https://github.com/atom/atom/blob/ae9f24602d80cb15fe0edc15c1aae37d44ee786c/src/pane.coffee#L770-L771). You mentioned that you were able to remove the last pane in a dock. Can you offer any guidance for how I could reproduce that scenario?

**Comment by jasonrudolph** [comment]
d599b52d36bb3a91c28450501e3c85c2d36a2085 adds failing tests to demonstrate the bug in #14442:

```
$ apm test
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------...FFFF........-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Dock
  activating the next pane
    when the dock has more than one pane
      it activates the next pane
        TypeError: dock.activateNextPane is not a function
          TypeError: dock.activateNextPane is not a function
          at /Users/j/github/atom/spec/dock-spec.js:78:14
          at .<anonymous> (/Users/j/github/atom/spec/async-spec-helpers.js:24:22)
    when the dock has only one pane
      it leaves the current pane active
        TypeError: dock.activateNextPane is not a function
          TypeError: dock.activateNextPane is not a function
          at /Users/j/github/atom/spec/dock-spec.js:93:14
          at .<anonymous> (/Users/j/github/atom/spec/async-spec-helpers.js:24:22)
  activating the previous pane
    when the dock has more than one pane
      it activates the previous pane
        TypeError: dock.activatePreviousPane is not a function
          TypeError: dock.activatePreviousPane is not a function
          at /Users/j/github/atom/spec/dock-spec.js:112:14
          at .<anonymous> (/Users/j/github/atom/spec/async-spec-helpers.js:24:22)
    when the dock has only one pane
      it leaves the current pane active
        TypeError: dock.activatePreviousPane is not a function
          TypeError: dock.activatePreviousPane is not a function
          at /Users/j/github/atom/spec/dock-spec.js:127:14
          at .<anonymous> (/Users/j/github/atom/spec/async-spec-helpers.js:24:22)


Finished in 1.261 seconds
15 tests, 52 assertions, 4 failures, 2120 skipped

Tests failed
```

ae9f24602d80cb15fe0edc15c1aae37d44ee786c then fixes those failures:

```
$ apm test
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------...............-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Finished in 1.288 seconds
15 tests, 56 assertions, 0 failures, 2120 skipped

Tests passed
```

File: spec/dock-spec.js (hunk 1 above)
Comment (lee-dohm) [inline spec/dock-spec.js]: I can't find a way to have zero panes in a dock :+1: I guess I was misremembering ¯\\\_(ツ)_/¯

File: spec/workspace-spec.js [hunk 2]
```diff
@@ -2528,6 +2528,62 @@ i = /test/; #FIXME\
     })
   })
 
+  describe('::activateNextPane', () => {
+    describe('when the active workspace pane is inside a dock', () => {
```
Comment (nathansobo) [inline spec/workspace-spec.js]: This looks great. Tiny nitpick that doesn't matter that much... I think you could just say "pane" instead of "workspace pane"

Comment (jasonrudolph) [inline spec/workspace-spec.js]: Thanks, @nathansobo.

> I think you could just say "pane" instead of "workspace pane"

I thought about that. I opted for the more verbose language ("workspace pane") to clarify that  we're talking about the workspace's active pane, as opposed to the active pane inside a specific pane container like the dock or workspace center. IIRC, each dock and the workspace center will have an active pane, but only one of those panes will be the workspace's active pane.
